src/simulator/tests/halo_simulator_test.py

```diff
@@ -90,6 +90,93 @@ def test_simulated_reach_by_spend_with_privacy(self, mock_laplace_mechanism):
         )
         self.assertTrue(reach_point.reach(1) >= 0)
 
+    @parameterized.named_parameters(
+        # testcase_name, num_publishers, spends, regions, expected
+        {
+            "testcase_name": "with_empty_regions",
+            "num_publishers": 2,
+            "spends": [0.005, 0.01],
+            "regions": {},
+            "expected": [
+                ReachPoint([0, 0], [0], [0.005, 0]),
+                ReachPoint([0, 0], [0], [0, 0.005]),
+                ReachPoint([0, 0], [0], [0.005, 0.005]),
+            ],
+        },
+        {
+            "testcase_name": "without_active_publishers",
+            "num_publishers": 2,
+            "spends": [0, 0],
+            "regions": {},
+            "expected": [],
+        },
+        {
+            "testcase_name": "with_1_active_pub_from_2_pubs",
+            "num_publishers": 2,
+            "spends": [0.04, 0.02],
+            "regions": {1: [2]},
+            "expected": [
+                ReachPoint([3, 0], [2], [0.04, 0]),
```

Comment:
The impressions list [3, 0] says that 3 impressions were shown.  But the k+ reach vector of [2] says that only 2 impressions were shown.  One of these two numbers must be incorrect.

Reply:
I think they both are correct. Please see the explanation below:
The impressions list shows that given the spends vector = [0.04, 0], ie. only publisher_1, we have 3 impressions for publisher_1 -- (1, 0.01), (2, 0.02), (1, 0.04).
And the k+ reaches vector indicates that there are only 2 unique users that were reached by publisher1 -- user_1 and user_2.

Reply:
You're right, thanks for correcting me on this.